### PR TITLE
[flang] Lower passing non assumed-rank/size to assumed-ranks

### PR DESCRIPTION
**Pull request by jeanPerier**
Start implementing assumed-rank support as described in https://github.com/llvm/llvm-project/blob/main/flang/docs/AssumedRank.md

This commit holds the minimal support for lowering calls to procedure with assumed-rank arguments where the procedure implementation is done in C.

The case for passing assumed-size to assumed-rank is left TODO since it will be done a change in assumed-size lowering that is better done in another patch.

Implementation of Fortran procedure with assumed-rank arguments is still TODO.

### CODE REVIEW

File: flang/lib/Lower/ConvertCall.cpp [hunk 1]
```diff
@@ -1043,39 +1083,55 @@ static PreparedDummyArgument preparePresentUserCallActualArgument(
...
+    if (baseBoxDummy.isAssumedRank())
+      if (const Fortran::semantics::Symbol *sym =
+              Fortran::evaluate::UnwrapWholeSymbolDataRef(expr))
+        if (Fortran::semantics::IsAssumedSizeArray(sym->GetUltimate()))
+          TODO(loc, "passing assumed-size to assumed-rank array");
+
+  // For ranked actual passed to assumed-rank dummy, the cast to assumed-rank
+  // box is inserted when building the fir.call op. Inserting it here would
+  // cause the fir.if results to be assumed-rank in case of OPTIONAL dummy,
+  // causing extra runtime costs due to the unknwon runtime size of assumed-rank
```
Comment (clementval) [inline flang/lib/Lower/ConvertCall.cpp:1131]:
```suggestion
  // causing extra runtime costs due to the unknown runtime size of assumed-rank
```

**Review by clementval — APPROVED** [review]
LGTM

File: flang/lib/Lower/ConvertCall.cpp [hunk 2]
```diff
@@ -942,6 +964,18 @@ static PreparedDummyArgument preparePresentUserCallActualArgument(
       (passingPolymorphicToNonPolymorphic ||
        !Fortran::evaluate::IsSimplyContiguous(expr, foldingContext));
 
+  // Create dummy type with actual argument rank when the dummy is an assumed
+  // rank. That way, all the operation to create dummy descriptors are ranked if
+  // the dummy is ranked, which allows simple code generation.
```
Comment (vzakhari) [inline flang/lib/Lower/ConvertCall.cpp:969]: I am not sure I am following the comment.  "if the dummy is ranked" seems to contradict the previous condition that "dummy is an assumed rank".  Can you please clarify?

Comment (jeanPerier) [reply flang/lib/Lower/ConvertCall.cpp:969]: Indeed, thanks for reading it! updated to "to create dummy descriptors are ranked if the actual argument is ranked"

Comment (jeanPerier) [inline flang/lib/Lower/ConvertCall.cpp:969]:
```suggestion
  // the actual argument is ranked, which allows simple code generation.
```

**Review by vzakhari — APPROVED** [review]

**Pull request merged**